### PR TITLE
Create and use MoyaError type for SignalProducers.

### PR DESCRIPTION
**Pull request by tomburns**
I took a blind stab at the change I was talking about in #288, mostly as an exercise for myself. I think this preserves the data presented previously in the NSErrors' userInfo, while providing a more swift-friendly interface for retrieving it.

As I said in the issue, I have no idea what I'm doing, but I'd love to hear your thoughts!

### CODE REVIEW

**Comment by tomburns** [comment]
Hmm, tests are passing for me locally after that last commit. Let me know if there's something I can do to make Travis happy :)

**Comment by ashfurrow** [comment]
Travis is failing for Moya intermittently (see: https://github.com/Moya/Moya/issues/259 ). I've restarted the build to see what happens, and I'll take a look at the code now.

File: Demo/DemoTests/SignalProducer+MoyaSpec.swift [hunk 1]
```diff
@@ -183,7 +183,7 @@ class SignalProducerMoyaSpec: QuickSpec {
                 }
                 
                 expect(receivedError).toNot(beNil())
-                expect(receivedError?.domain).to(equal("\(NSCocoaErrorDomain)"))
+//                expect(receivedError?.domain).to(equal("\(NSCocoaErrorDomain)"))
```
Comment (tomburns) [inline Demo/DemoTests/SignalProducer+MoyaSpec.swift]: Forgot to comment earlier: I wasn't clear on a good replacement for this line

**Comment by ashfurrow** [comment]
Yup, green now :smile: :green_apple: 

This looks great, fantastic work :100:

The `MoyaErrorCode` enum exists, but I feel like (given the direction Swift 2 has taken with error-handling) it may be time to retire it altogether and use just your new `MoyaError` enum. This would only change using Moya with ReactiveCocoa (maybe we could make corresponding changes for RxSwift). Any opinions on that? /cc @Moya/contributors 

There're a few tests I'd want to make sure we cover in `SignalProducer+MoyaSpec.swift` but we can sort those out later.

Updating the changeling is a good idea in the mean time – right at the top under "Next". Make sure to mark them with **bold breaking changes** and give yourself credit. Thanks again for the awesome pull request – we'll discuss it for a bit and get your changes into a release hopefully this week. Cheers! :coffee:

File: Demo/DemoTests/SignalProducer+MoyaSpec.swift (hunk 1 above)
Comment (ashfurrow) [inline Demo/DemoTests/SignalProducer+MoyaSpec.swift]: Hmm, good question. I've not compared enum cases in Nimble matchers before. I looks like [a question others have had](https://github.com/Quick/Nimble/issues/212). If you want, take a stab at one of the solutions @jeffh suggests there – I can lend a hand to get these working, too.

**Comment by petester42** [comment]
Im all for moving to a typed error instead of NSError. We could possibly use the typed error everywhere instead of just the reactive extensions. It would make things more consistent. The same could be done for `MoyaResponse` as I discussed [here](https://github.com/Moya/Moya/issues/198#issuecomment-154258823)

**Comment by tomburns** [comment]
Pushed a simple comparison function for the error type; I wasn't comfortable shoving that type comparison into `Equatable` without further discussion. Also updated the changelog.

**Comment by tomburns** [comment]
While we're looking at this change, it might be worth examining which associated values we'd like to have for each error. As an example: I've already added the status code as an `Int`, but it might be nice to have the expected range which resulted in the error as well? e.g. `.StatusCode(404,200...299,response)`

File: Source/ReactiveCore/MoyaError.swift [hunk 2]
```diff
@@ -9,3 +9,59 @@ public enum MoyaErrorCode: Int {
...
+
+public enum MoyaError: ErrorType {
+    case ImageMapping(MoyaResponse)
+    case JSONMapping(MoyaResponse)
+    case StringMapping(MoyaResponse)
+    case StatusCode(Int,MoyaResponse)
+    case Data(MoyaResponse)
+    case Underlying(ErrorType)
+    
+    public func sameErrorType(otherError: MoyaError) -> Bool {
```
Comment (ashfurrow) [inline Source/ReactiveCore/MoyaError.swift]: Where there were a better way to structure this function, but I don't think there is. Is it useful to developers using Moya, or just us when testing? Either way, I think it might look cleaner inside an extension.

**Comment by ashfurrow** [comment]
Agreed it's worth being thoughtful about their associated values. I'm not sure how I feel about the range specification. I _think_ I like it – anyone else have feels?

File: Source/ReactiveCore/MoyaError.swift (hunk 2 above)
Comment (petester42) [inline Source/ReactiveCore/MoyaError.swift]: Making the error conform to Equatable would be a better solution in my opinion.

Comment (tomburns) [inline Source/ReactiveCore/MoyaError.swift]: This isn't equality per se though; to my mind that would involve the associated values matching as well.

Moving it to an extension makes sense; I'm trying to come up with non-testing use cases but nothing so far. Maybe we move it into the test target for now and see if we miss it?

**Comment by tomburns** [comment]
To keep the associated values clean (especially given their lack of names) we could also consider specialized struts, like so:
```
public enum MoyaError: ErrorType {
    case StatusCode(StatusCodeErrorMetadata)
    [...]
}

struct StatusCodeErrorMetadata {
    let encounteredStatusCode: Int
    let expectedStatusRange: Range<Int>
    let response: MoyaResponse
}
```

That struct's name is a bit of a mouthful, but I like that it keeps us down to one associated value and the convention is fairly straightforward.

**Comment by petester42** [comment]
You mentioned that you want to keep associated values clean "especially given their lack of names". You can give names to associated values inside an enum in swift if that's what you wanted to do. So like `case CoolCase(value: Sting)`

**Comment by justinmakaila** [comment]
I think associated values are the way to go. They represent a domain, and each case can encapsulate any values associated with it.



—Justin Makaila

**Comment by tomburns** [comment]
Ah, glad to learn that names work! No idea why I was so sure they didn't. 

+1 for associated values, then :joy:

File: Source/ReactiveCore/MoyaError.swift (hunk 2 above)
Comment (ashfurrow) [inline Source/ReactiveCore/MoyaError.swift]: Agreed, this function works great. Moving it to the test target sounds great too, we can always move it back later.

**Comment by tomburns** [comment]
I broke it again! :tada: 

Tests are passing locally for me; not sure what's up: 

```
tburns@Monarch ~/code/Moya [error-type *]
± % rake test                                                                                                                              !10072
set -o pipefail && xcodebuild -workspace 'Demo.xcworkspace' -scheme 'Demo' -configuration 'Debug' -sdk iphonesimulator -destination platform='iOS Simulator',OS='9.0',name='iPhone 6' build test | xcpretty -c --test
2015-11-08 23:02:13.836 xcodebuild[73178:2107923] [MT] IDETestOperationsObserverDebug: (FC2F7820-8995-42A0-8796-B9ADFD7D4C44) Beginning test session FC2F7820-8995-42A0-8796-B9ADFD7D4C44 with Xcode 7B91b on target <DVTiPhoneSimulator: 0x7f9b2d6ef3f0> {
		SimDevice: SimDevice : iPhone 6 (C5B3EBC8-B35E-44DE-8D89-72DCB622959D) : state={ Shutdown } deviceType={ SimDeviceType : com.apple.CoreSimulator.SimDeviceType.iPhone-6 } runtime={ SimRuntime : 9.0 (13A344) - com.apple.CoreSimulator.SimRuntime.iOS-9-0 }
} (9.0 (13A344))
..................................................................

	 Executed 66 tests, with 0 failures (0 unexpected) in 5.932 (5.959) seconds
```

**Comment by petester42** [comment]
You didn't add the `MoyaError+MoyaSpec.swift` to the DemoTests target inside the demo project. Should work after you do that.

**Comment by JRHeaton** [comment]
Very cool, I think this is a great way to handle errors 👍. 

A couple of thoughts:
1. With Swift's sweet pattern matching, I think you may able to trim some redundant code off of the `sameErrorType` function by switching on a tuple of the two values. Example:
```swift
switch (self, otherError) {
case (.ImageMapping, .ImageMapping), (.JSONMapping, .JSONMapping) /*, etc... */:
    return true
default: 
    return false
}
```
2. Since Alamofire fails with `NSError`, I think it may make sense to type the associated value on `Underlying` to be `NSError` rather than `ErrorType`. This prevents confusion that there could be any other type given and thus need to do runtime identification of the type that we know at compile time. This would require re-typing `EndpointSampleResponse.NetworkError`'s associated error type as well, so that `Moya.Completion` could have it's error type be `NSError`.

In short, the Alamofire/"underlying" layer only fails with `NSError`, and since none of these status code filtration or data mapping functions exist in the non-reactive Moya core, it is the only error type to be dealt with, so we can benefit from some type specificity here.

**Comment by tomburns** [comment]
@JRHeaton The only nice thing about the current shape of that comparison function is that it's exhaustive, and will generate warnings if we add to the enum. Your version is certainly the more handsome of the two; what does the group think?

As an aside, the latest CI failure appears to be Travis flapping again.

**Comment by petester42** [comment]
I like pattern matching more but that's just me. Also the switch can be exhaustive without default if that is what you are worried about. You just need to add all the combinations as cases. I don't think having a default case that returns false would be bad. Usually we would want to check that 2 errors are the same so if there is a new error and that function has not been updated then it will return false so tests would fail etc.

**Comment by gregpardo** [comment]
I am also on board with moving away from NSError pattern. Can we just put the NSError into the metadata in case it's needed as well StatusCodeErrorMetadata @tomburns suggested?

**Comment by JRHeaton** [comment]
Good point @tomburns, it's a bit of concision versus compile-time safety. I wish there was a prettier way to get the latter, but oh well.

Any thoughts on my first point above? Alamofire only uses `NSError`, so the `MoyaError` case that covers the domain of Alamofire errors should provide an associated `NSError`, not `ErrorType`.

**Comment by ashfurrow** [comment]
I personally like ErrorType – it's convertible to NSError.

**Comment by JRHeaton** [comment]
Just seems silly to declare things as ErrorType when we _know_ it's literally only _ever_ an NSError. Am I not understanding a benefit of doing it the other way? I'm not so much complaining as I am truly not understanding, I guess. This is even the motivation behind parametric error types for ReactiveCocoa signals – to not leave a signal that can throw deterministic error types to the open-ended `ErrorType` that could be anything, but to give a specific `ErrorType`-conforming type so that there is no nondeterminism about types at compile time. It _is_ convertible to NSError, but it also ALREADY is an NSError, hence my not understanding declaring it differently. And this only applies of course to the `Underlying` error case.

Think about the call point of the Alamofire request. The completion handler provided takes an NSError, which Moya gets back, and then (only ever) feeds as `Underlying`, which takes a more generic type than we know.

**Comment by colinta** [comment]
Is there anything preventing Alamofire from converting to using `ErrorType` over `NSError`?  If they do that, Moya would have to push out a new version.  Not *at all* a bad thing, and this is the kind of version bump that makes semantic versioning so relevant, but that *might* be an argument in favor of the more generic `ErrorType`.

A way of justifying it philosophically might be: if we don't need / aren't consuming the features of `NSError`, then using the generic type indicates that we *aren't promising* what type of error is coming down the wire.  If this error is coming from a dependency, then this lack of promise is honest - we know we'll get an error, but we can't promise what it will be.

I'm just catching up on this discussion, but these are my initial thoughts.  Gonna re-read now.

**Comment by ashfurrow** [comment]
The ability to write exhaustive catch blocks is actually a really good point. 

Let's try to remember that we're all after the same thing (writing great code) even if we disagree on how to get there. If it would help, I can schedule a Google Hangout to discuss some of the pros and cons.

**Comment by gregpardo** [comment]
Should an issue be opened on Alamofire? I do believe it would be an API breaking change but I think it's probably worth pursuing. The docs on swift errors suggest using the ErrorType pattern is the way to go.

**Comment by ashfurrow** [comment]
We should look into the history of their error handling first. The changelog might reveal why they have it this way.

**Comment by gregpardo** [comment]
Coincidentally, their 'benefits to upgrading' to 3.0 is almost all about errors.

I think the part that may be of interest is the custom wrapping of any NSError to a type of our choosing. 

https://github.com/Alamofire/Alamofire/blob/0cb96d59ad47be1b6f5018188fcfd1d6b0ecbb0c/Documentation/Alamofire%203.0%20Migration%20Guide.md#benefits-of-upgrading

**Comment by ashfurrow** [comment]
Cool, not sure where we stand on this one – anything stopping it from being merged? 

In the mean time, I'm going to merge https://github.com/Moya/Moya/pull/291 because it has no breaking changes, and it'd be cool to include its changes in a point release before this one (so people can have just those changes if they want).

**Comment by tomburns** [comment]
I think this can be merged as a first step, unless we want to go all-in on the new error type all at once. If there are any pending changes I've missed, let me know and i'll try to address them.

**Comment by tomburns** [comment]
Scanning the thread, the only things that I see as unresolved are what associated values we want to make available on each error, and naming them inside the tuples. Do we want to take care of that before merging, or just before releasing this into the wild?

**Comment by ashfurrow** [comment]
Cooooool yeah sounds great! There've been changes on master since this PR was submitted, @tomburns could I ask you to merge from master?

**Comment by tomburns** [comment]
Merge to master: :white_check_mark: 

If someone could give that a quick spot check I think we're good to go.

**Comment by petester42** [comment]
After those small changes I think it should be good. It might cause you to have to update other parts of the codebase like the tests and the RxSwift extension though. If you need help with that, I'm here to help. :+1:

**Comment by tomburns** [comment]
Thanks for the feedback! Good catches all around; I agree with all the proposed changes and I'll try to get this updated today.

**Comment by tomburns** [comment]
I believe this is ready for further review; as it stands I was able to make the suggested changes without touching the Rx extensions or tests. I'm a little unhappy with that revised attempt at passing along the status code error (see my comment on the latest commit above) , so if anyone has any tips I'm all ears.

**Comment by ashfurrow** [comment]
Cool, seems good to me. If someone else could give this a final one-over and merge, that'd be :100: Good work everyone :balloon:

**Comment by petester42** [comment]
Looks good. There might be a little clean up needed to maybe remove `MoyaErrorDomain` and `MoyaErrorCode` if possible. If no one has other objections I will merge this.
